### PR TITLE
Improve handling of imaginary datetimes

### LINKED ISSUE #72: Wrong result when a shift involves a DST change.
In paris in 2013, we changed to DST in 2013-03-31 (at 2 AM it was 3 AM), from UTC+1 to UTC+2 (DST)

Here's an example of the issue I have with arrow : 

``` python
>>> import arrow, datetime

>>> just_before = arrow.get(datetime.datetime(2013, 03, 31, 1, 50, 45), "Europe/Paris").ceil("hour")
>>> just_before
<Arrow [2013-03-31T01:59:59.999999+01:00]>

>>> just_after = just_before.replace(microseconds=+1)
>>> just_after
<Arrow [2013-03-31T02:00:00+02:00]>
>>> # That is not right... It should be either 2 AM UTC+1 or preferably  3AM UTC +2
>>> # One could argue that depending on what you ask for, it may be the correct answer, but then the following is :

>>> (just_after - just_before).total_seconds()
1e-06
>>> # That is right but not consistent with the dates

>>> (just_after.to("utc") - just_before.to("utc").total_seconds()
-3599.999999
# That should be the same value as the previous computation. Plus, the value is negative because :

>>> just_before.to('utc'), just_after.to('utc')
(<Arrow [2013-03-31T00:59:59.999999+00:00]>,
 <Arrow [2013-03-31T00:00:00+00:00]>)
```

I think the problem is pretty clear...

### PR DESCRIPTION
**Pull request by systemcatch**
## Pull Request Checklist

Thank you for taking the time to improve Arrow! Before submitting your pull request, please check all *appropriate* boxes:

- [x] 🧪 Added **tests** for changed code.
- [x] 🛠️ All tests **pass** when run locally (run `tox` or `make test` to find out!).
- [x] 🧹 All linting checks **pass** when run locally (run `tox -e lint` or `make lint` to find out!).
- [ ] 📚 Updated **documentation** for changed code.
- [x] ⏩ Code is **up-to-date** with the `master` branch.

If you have *any* questions about your code changes or any of the points above, please submit your questions along with the pull request and we will try our best to help!

## Description of Changes

closes #72 

Pendulum example is now fixed.

```shell
(arrow) chris@ThinkPad:~/arrow$ python
Python 3.8.3 (default, Jul  7 2020, 18:57:36) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import arrow
>>> just_before = arrow.get(2013, 3, 31, 1, 55, "Europe/Paris")
>>> just_before = arrow.get(2013, 3, 31, 1, 55, tzinfo= "Europe/Paris")
>>> just_before
<Arrow [2013-03-31T01:55:00+01:00]>
>>> just_before.shift(minutes=+10)
<Arrow [2013-03-31T03:05:00+02:00]>
```

Range duplication no longer occurs.
```shell
>>> before = arrow.get('2018-03-10 23:00:00', 'YYYY-MM-DD HH:mm:ss', tzinfo='US/Pacific')
>>> after = arrow.get('2018-03-11 04:00:00', 'YYYY-MM-DD HH:mm:ss', tzinfo='US/Pacific')
>>> result=[ (t, t.to('utc')) for t in arrow.Arrow.range('hour', before, after) ]
>>> for r in result:
...     print(r)
... 
(<Arrow [2018-03-10T23:00:00-08:00]>, <Arrow [2018-03-11T07:00:00+00:00]>)
(<Arrow [2018-03-11T00:00:00-08:00]>, <Arrow [2018-03-11T08:00:00+00:00]>)
(<Arrow [2018-03-11T01:00:00-08:00]>, <Arrow [2018-03-11T09:00:00+00:00]>)
(<Arrow [2018-03-11T03:00:00-07:00]>, <Arrow [2018-03-11T10:00:00+00:00]>)
(<Arrow [2018-03-11T04:00:00-07:00]>, <Arrow [2018-03-11T11:00:00+00:00]>)
```

I was also considering wrapping dateutil's `resolve_imaginary` function as an utility method for our users.

### CODE REVIEW

File: tests/test_arrow.py [hunk 1]
```diff
@@ -838,6 +847,82 @@ def test_shift_quarters_bug(self):
...
+        assert new_york.shift(hours=-2) == arrow.Arrow(
+            2011, 3, 13, 1, 30, tzinfo="America/New_York"
+        )
+
+        london = arrow.Arrow(2019, 3, 31, 2, tzinfo="Europe/London")
+        assert london.shift(hours=-1) == arrow.Arrow(
+            2019, 3, 31, 2, tzinfo="Europe/London"
+        )
+        assert london.shift(hours=-2) == arrow.Arrow(
+            2019, 3, 31, tzinfo="Europe/London"
```
Comment (jadchaar) [inline tests/test_arrow.py:896]: Should we explicitly place the hour=0 here for clarity?

**Comment by systemcatch** [comment]
> First pass looks solid. Is this seriously all the places where we need to cover imaginary dates/times?

I think so, by fixing `shift` we can use it in `range` and `span` which then propagate to other methods.

> Does this address https://github.com/arrow-py/arrow/issues/686 as well? If so, we should add a regression test for it.

No and looking at it I don't think that issue can be fixed in arrow. Those comparison results are baked into datetime itself.

**Comment by systemcatch** [comment]
@pganssle any thoughts to add here?

File: arrow/arrow.py [hunk 2]
```diff
@@ -836,6 +843,9 @@ def shift(self, **kwargs):
 
         current = self._datetime + relativedelta(**relative_kwargs)
```
Comment (jadchaar) [inline arrow/arrow.py:844]: Do we need to call shift instead on all uses of `relativedelta`?

File: arrow/arrow.py [hunk 3]
```diff
@@ -722,7 +722,14 @@ def ambiguous(self):
 
         return dateutil_tz.datetime_ambiguous(self._datetime)
 
-    # mutation and duplication
+    @property
+    def imaginary(self):
+        """Returns a boolean indicating whether the :class: `Arrow <arrow.arrow.Arrow>` object exists in the current
+        timezone"""
```
Comment (jadchaar) [inline arrow/arrow.py:728]: I'd compress this onto one line and add a period: Indicates whether the :class: `Arrow <arrow.arrow.Arrow>` object exists in the current timezone.

**Review by jadchaar — APPROVED** [review]
Looks good other than a single docstring comment.

Should we also resolve imaginary datetimes if they are constructed via the `arrow.get` or `arrow.Arrow`?

**Comment by systemcatch** [comment]
> Should we also resolve imaginary datetimes if they are constructed via the `arrow.get` or `arrow.Arrow`?

I tend to think not as `datetime` lets you construct imaginary instances directly. Users might have a reason for needing them and with the new `imaginary` property it's easy to check it

**Comment by systemcatch** [comment]
Hey @jadchaar I've made the change you suggested, let me know if you're happy and I will merge this in.

**Comment by jadchaar** [comment]
Looks great! Ship it!

**Pull request merged**